server/src/main/java/org/elasticsearch/action/search/SearchResponse.java

```diff
@@ -507,53 +518,99 @@ public Clusters(int total, int successful, int skipped, int remoteClusters, bool
             this.total = total;
             this.successful = successful;
             this.skipped = skipped;
-            this.remoteClusters = remoteClusters;
             this.ccsMinimizeRoundtrips = ccsMinimizeRoundtrips;
+            this.clusterInfo = new ConcurrentHashMap<>();
         }
 
         /**
-         * Assumes ccsMinimizeRoundtrips=false.
-         * We are not tracking number of remote clusters in this search.
+         * For use with cross-cluster searches.
+         * When minimizing roundtrips, the number of successful and skipped clusters is not known until
+         * the end of the search and it the information in SearchResponse.Cluster object will be updated
+         * as each cluster returns.
+         * @param localIndices The localIndices to be searched - null if no local indices are to be searched
+         * @param remoteClusterIndices mapping of clusterAlias -> OriginalIndices for each remote cluster
+         * @param ccsMinimizeRoundtrips whether minimizing roundtrips for the CCS
          */
-        public Clusters(int total, int successful, int skipped) {
-            this(total, successful, skipped, true);
+        public Clusters(
```

Comment:
There are two separate ways to use `SearchResponse.Clusters`.

The original way is that `SearchResponse.Clusters` is immutable and just holds counters for total, successful, and skipped. A typical usage is just to use SearchResponse.Clusters.EMPTY when you don't even want to track that accounting (local only searches). For CCS MRT=false, an immutable SearchResponse.Clusters is created after initially contacting each remote cluster (in order to do can-match via SearchShards API) and the Clusters object is never updated after that (even if one of the remote clusters later fails during the search).

For CCS MRT=true in this PR, the Clusters object is now composed of several underlying `SearchResponse.Cluster` object, each of which is updated separately as the remote clusters report back. So in this model an initial Clusters object is created at the beginning of the search and it is updated periodically from separate callback threads.

This model can also be used in a future PR for CCS MRT=false, but that is out of scope for this PR.